Mathlib/Topology/CompletelyRegular.lean

```diff
@@ -0,0 +1,93 @@
+/-
+Copyright (c) 2032 Matias Heikkilä. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Matias Heikkilä
+-/
+import Mathlib.Topology.UrysohnsLemma
+import Mathlib.Topology.UnitInterval
+
+/-!
+# Completely regular topological spaces.
+
+This file defines `CompletelyRegularSpace`.
+
+## Main definitions
+
+* `CompletelyRegularSpace`: A completely regular space `X` is a T₁ space where, for each
+  closed set `K ⊆ X` and a point `x ∈ Kᶜ`, there is a continuous function `f` from
+  `X` to the unit interval, so that `f x = 0` and `f k = 1` for all `k ∈ K`. A completely
+  regular space is a T₂ space and a T₄ space is completely regular.
+
+## Main results
+
+### Completely regular spaces
+
+* `CompletelyRegularSpace`: Given a point `x` and a closed set `K` such that `{x} ∩ K = ∅ `,
+  we can find a continuous real-valued function `f` such that `f` is equal to `0` on `{x}`
+  and `f` is equal to `1` on `K`.
+* `CompletelyRegularSpace.t3space`: A completely regular space is a T₃ space.
+* `NormalSpace.completelyRegularSpace`: A normal space is a completely regular space.
+
+## Implementation notes
+
+The present definition `CompletelyRegularSpace` is a slight modification of the one given in
+[russell1974]. There it's assumed that any point `x ∈ Kᶜ` is separated from the closed set `K` by a
+continuous *real* valued function `f` (as opposed to `f` being unit-interval-valued). This can be
+converted to the present definition by replacing a real-valued `f` by `h ∘ g ∘ f`, with
+`g : x ↦ max(x, 0)` and `h : x ↦ min(x, 1)`.
+
+## References
+
+* [Russell C. Walker, *The Stone-Čech Compactification*][russell1974]
+-/
+
+universe u
+
+noncomputable section
+
+open Set Topology unitInterval
+
+variable {X : Type u} [TopologicalSpace X] [T1Space X]
+
+/-- A T1 space X is completely regular if points can be separated from closed sets via
+  continuous functions to the unit interval. -/
+@[mk_iff completelyRegularSpace_iff]
+class CompletelyRegularSpace (X : Type u) [TopologicalSpace X] [T1Space X] : Prop where
+  completely_regular : ∀ (x : X), ∀ (K : Set X) (_: IsClosed K), x ∉ K →
+    ∃ f : X → I, Continuous f ∧ f x = 0 ∧ EqOn f 1 K
```

Comment:
I think you should remove `T1Space` from the definition, because e.g. on [Wikipedia](https://en.wikipedia.org/wiki/Tychonoff_space) a completely regular space is not necessarily T1; if it's T0 then it's T3 and called a Tychonoff space. The equivalence with uniformizability only holds without the T1 condition, since a uniform space isn't necessarily T1.

Also I think you can use `C(X, I)` rather than separating out the `Continuous` condition. I know that @PatrickMassot urged you to separate it out, but I don't see the point: if you're getting out a continuous map from a `CompletelyRegularSpace`, then working with a bundled continuous map will only be convenient; if you're proving a `CompletelyRegularSpace`, then writing `⟨⟨f, cont_f⟩, ..⟩` instead of `⟨f, cont_f, ...⟩` isn't really much more difficult. I can see that an unbundled `Continuous` condition would be more convenient if it appears as an argument to a theorem, but not here. Note also that [Urysohn's lemma](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Topology/UrysohnsLemma.html#exists_continuous_zero_one_of_closed) is stated using bundled ContinuousMap.

Reply:
About the terminology: I went back and forth a bit before making this PR, but eventually decided to use the definition given in Walker's book. Reconsidering this, I realize that some of the things I'm planning to add do not require the T1 bit, so it could be better to exclude T1 here, and maybe add a definition for another type of space (`Tychonoff`??) with T1 included.

About the bundled functions: I don't really have the sufficient background in contributing to mathlib to add any input on that.

Reply:
> it could be better to exclude T1 here, and maybe add a definition for another type of space (TychonoffSpace??) with T1 included.

Yeah it's definitely reasonable, e.g. [T3Space](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Topology/Separation.html#T3Space) is T0 + RegularSpace. So CompleteRegularSpace without T0 is consistent with that, and CompleteRegularSpace + T0 could be called T35Space in analogy with [T25Space](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Topology/Separation.html#T25Space).

Reply:
I just checked in Mathlib and *every* use of Urysohn's lemma unpack the bundle continuous map. So thanks @alreadydone for pointing out we should restate that lemma without bundling. I think it is not worth the trouble to fix Urysohnś lemma, but let's not reproduce that mistake here.

Reply:
Excuse me? I don't see it being unpacked at all in Metrizable, CompHaus/Basic, ContinuousFunction/Ideals, ContinuousMapDense among the [search results](https://github.com/search?q=repo%3Aleanprover-community%2Fmathlib4%20exists_continuous_zero_one_of_closed&type=code). Why would it be unpacked? ContinuousMap has a FunLike instance so if you want the function you can just pretend it's a function, and for the `Continuous` condition you can just write `f.continuous` or even `f.2`.

Reply:
Oh I see my mistake. I saw huge unpacking in lines like `let ⟨f, hfs, hft, hf⟩ := exists_continuous_zero_one_of_closed hs ht hd`, but you are right that the bundle continuous map is not unpacked. So there is indeed no need to fix those (but the also bring nothing).